homeassistant/components/remote/xiaomi_miio.py

```diff
@@ -0,0 +1,254 @@
+from homeassistant.const import (
+    CONF_NAME, CONF_HOST, CONF_TOKEN, CONF_TIMEOUT,
+    ATTR_ENTITY_ID, ATTR_HIDDEN, CONF_COMMAND)
+import homeassistant.helpers.config_validation as cv
+from homeassistant.util.dt import utcnow
+
+REQUIREMENTS = ['python-miio==0.3.4']
+
+_LOGGER = logging.getLogger(__name__)
+
+SERVICE_LEARN = 'xiaomi_miio_learn_command'
+PLATFORM = 'xiaomi_miio'
+
+CONF_SLOT = 'slot'
+CONF_COMMANDS = 'commands'
+
+DEFAULT_TIMEOUT = 10
+DEFAULT_SLOT = 1
+
+LEARN_COMMAND_SCHEMA = vol.Schema({
+    vol.Required(ATTR_ENTITY_ID): vol.All(str),
+    vol.Optional(CONF_TIMEOUT, default=10):
+        vol.All(int, vol.Range(min=0)),
+    vol.Optional(CONF_SLOT, default=1):
+        vol.All(int, vol.Range(min=1, max=1000000)),
+})
+
+COMMAND_SCHEMA = vol.Schema({
+    vol.Required(CONF_COMMAND): vol.All(cv.ensure_list, [cv.string])
+    })
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Optional(CONF_NAME): cv.string,
+    vol.Required(CONF_HOST): cv.string,
+    vol.Optional(CONF_TIMEOUT, default=DEFAULT_TIMEOUT):
+        vol.All(int, vol.Range(min=0)),
+    vol.Optional(CONF_SLOT, default=DEFAULT_SLOT):
+        vol.All(int, vol.Range(min=1, max=1000000)),
+    vol.Optional(ATTR_HIDDEN, default=True): cv.boolean,
+    vol.Required(CONF_TOKEN): vol.All(str, vol.Length(min=32, max=32)),
+    vol.Optional(CONF_COMMANDS, default={}):
+        vol.Schema({cv.slug: COMMAND_SCHEMA}),
+}, extra=vol.ALLOW_EXTRA)
+
+
+@asyncio.coroutine
+def async_setup_platform(hass, config, async_add_devices, discovery_info=None):
+    """Set up the Xiaomi IR Remote (Chuangmi IR) platform."""
+    from miio import ChuangmiIr
+    from construct import ChecksumError
+
+    host = config.get(CONF_HOST)
+    token = config.get(CONF_TOKEN)
+
+    # Create handler
+    _LOGGER.info("Initializing with host %s (token %s...)", host, token[:5])
+    device = ChuangmiIr(host, token)
+
+    try:
+        device.info()
+
+    # This should be DeviceError but python-miio returns wrong except.
+    except ChecksumError as ex:
+        _LOGGER.error("Token not accepted by device : %s", ex)
+        return
+
+    if PLATFORM not in hass.data:
+        hass.data[PLATFORM] = {}
+
+    entity_id = config.get(CONF_NAME, "xiaomi_miio_" +
+                           host.replace('.', '_'))
+    slot = config.get(CONF_SLOT)
+    timeout = config.get(CONF_TIMEOUT)
+
+    hidden = config.get(ATTR_HIDDEN)
+
+    xiaomi_miio_remote = XiaomiMiioRemote(
+        entity_id, device, slot, timeout, hidden, config.get(CONF_COMMANDS))
+
+    hass.data[PLATFORM][host] = xiaomi_miio_remote
+
+    async_add_devices([xiaomi_miio_remote])
+
+    @asyncio.coroutine
+    def async_service_handler(service):
+        """Handle a learn command."""
+        if service.service != SERVICE_LEARN:
+            _LOGGER.error("We should not handle service: %s", service.service)
+            return
+
+        entity_id = service.data.get(ATTR_ENTITY_ID)
+
+        entity = None
```

Comment:
I think there's no need for a newline between these two, but it may be a matter a taste.